src/main/java/org/springframework/data/elasticsearch/repository/support/MappingElasticsearchEntityInformation.java

```diff
@@ -62,12 +59,12 @@ public String getIdAttribute() {
 
 	@Override
```

Comment:
so this change moves the resolution of the `indexName` and the `type` to the getter methods in case that the indexName and type passed in the constructor are null. But this constructor is called only from two places:

```
	public MappingElasticsearchEntityInformation(ElasticsearchPersistentEntity<T> entity) {
		this(entity, entity.getIndexName(), entity.getIndexType(), entity.getVersionType());
	}
```
and 
```
	private <T, ID> ElasticsearchEntityInformation<T, ID> getEntityInformation(Class<T> domainClass,
			@Nullable RepositoryInformation information) {

		ElasticsearchPersistentEntity<?> entity = mappingContext.getRequiredPersistentEntity(domainClass);

		return new MappingElasticsearchEntityInformation<>((ElasticsearchPersistentEntity<T>) entity, entity.getIndexName(),
				entity.getIndexType(), entity.getVersionType());
	}
```
the second being in the class `ReactiveElasticsearchRepositoryFactory`.

And the parameters passed into the constructor for `indexName` and `type`are the same values that will be calculated later, they are either not null, so the later evaluation will not take place, or they are null, then they will evaluate to null later as well.

So this does not really change the current behaviour, or am I missing something?

Reply:
The whole point of this PR is to allow `entity.getIndexName()` to return different values over time: if the entity declares a SpEL expression that uses a spring bean, the latter will be called once when the context starts, and then several times while the repository is used.

During context start, the SpEL may return null (which was failing because of the Asserts which this PR removes). Later it will return non null values to resolve to an actual index name.

Reply:
ok, got it